src/main/java/org/craftercms/deployer/impl/lifecycle/aws/DuplicateS3LifecycleHook.java

```diff
@@ -144,12 +139,12 @@ protected void doExecute(Target target) throws DeployerException {
             return;
         }
         List<String> paths = getItemPathList(repoPath);
-        AmazonS3 client = buildClient(builderConfigurer);
+        S3AsyncClient client = buildClient(builderConfigurer);
         try {
-            AwsUtils.copyObjects(client, threadPoolTaskExecutor::getThreadPoolExecutor, getBucket(srcS3Url, sourceSiteName), getS3BaseKey(srcS3Url, sourceSiteName),
+            AwsUtils.copyObjects(client, threadPoolTaskExecutor.getThreadPoolExecutor(), getBucket(srcS3Url, sourceSiteName), getS3BaseKey(srcS3Url, sourceSiteName),
```

Comment:
This is where it's useful to see if in V2 we now can copy folders from one bucket to another, to try to speed things up @phuongnq @jmendeza

Reply:
It seems we can achieve that via S3 Batch Ops
https://docs.aws.amazon.com/AmazonS3/latest/userguide/copy-object.html
```
Copying multiple objects in a single request
To copy more than one Amazon S3 object with a single request, you can also use S3 Batch Operations. You provide S3 Batch Operations with a list of objects to operate on. S3 Batch Operations calls the respective API operation to perform the specified operation. A single Batch Operations job can perform the specified operation on billions of objects containing exabytes of data.
```
https://docs.aws.amazon.com/AmazonS3/latest/userguide/batch-ops.html#batch-ops-basics

Reply:
@sumerjabri @avasquez614 @jmendeza We can use the batch ops but it may require some additional configuration such as specifying the Accound ID and a role ARN for the job. (Btw, I think this feature is not new in SDK v2, it has been there in v1: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3control/AWSS3Control.html#createJob-com.amazonaws.services.s3control.model.CreateJobRequest-)

I have reviewed this doc: 
* https://docs.aws.amazon.com/AmazonS3/latest/userguide/tutorial-s3-batchops-lambda-mediaconvert-video.html 
* And this one for a sample code in SDK v1: https://docs.aws.amazon.com/AmazonS3/latest/userguide/batch-ops-create-job.html

The steps would be:

* Create a manifest.csv file to list all object that needs to be copied with the format `source_bucket_name,source_key_path`
* Upload this file to a S3 bucket (for example, we can just use the destination bucket)
* Create a job to copy from the source to the destination bucket using the `manifest.csv` (The job will read from `manifest.csv` and perform the copy for us)

```
CreateJobRequest createJobRequest = CreateJobRequest.builder()
                .accountId("ACCOUNT_ID")
                .operation(jobOperation)
                .manifest(manifest)
                .report(jobReport)
                .priority(1)
                .description("Copy objects from " + sourceBucket + " to " + targetBucket)
                .roleArn("arn:aws:iam::ACCOUNT_ID:role/ROLE_NAME")
                .confirmationRequired(false)
                .clientRequestToken(UUID.randomUUID().toString())
                .build();
CreateJobResponse createJobResponse = s3ControlClient.createJob(createJobRequest);
``` 
* We will also need a `S3ControlClient` for this

Just wondering if we should move this to a new ticket to continue discussing the solution/configuration?

Other than this, I don't see the new S3TransferManager has a method to copy a folder from one bucket to another. It only has an example of copying a key: https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/transfer-manager.html#transfer-manager-copy

If it's about uploading, we can upload a folder to a bucket though: https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/transfer-manager.html#transfer-manager-upload_directory

Reply:
I agree this is not part of the upgrade to SDK v2. Let's discuss in a new ticket

Reply:
Agree